File: backend/src/main/java/com/NBE3_4_2_Team4/standard/util/test/ConcurrencyTestUtil.java
package com.NBE3_4_2_Team4.standard.util.test;

import org.springframework.transaction.annotation.Transactional;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Supplier;

public class ConcurrencyTestUtil {

    public static <T> void execute(int threadCount, Supplier<T> task) {

        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
        CountDownLatch latch = new CountDownLatch(threadCount);
        CyclicBarrier barrier = new CyclicBarrier(threadCount);

        for (int i = 0; i < threadCount; i++) {
            executorService.submit(() -> {
                try {
                    barrier.await();
                    task.get();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                } finally {
                    latch.countDown();
                }
            });
        }

        try {
        latch.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        executorService.shutdown();
    }
}

File: backend/src/main/java/com/NBE3_4_2_Team4/standard/util/test/ConcurrencyTestUtil.kt
package com.NBE3_4_2_Team4.standard.util.test

import java.util.concurrent.CountDownLatch
import java.util.concurrent.CyclicBarrier
import java.util.concurrent.Executors
import java.util.function.Supplier

object ConcurrencyTestUtil {
    @JvmStatic
    fun <T> execute(threadCount: Int, task: Supplier<T>) {
        val executorService = Executors.newFixedThreadPool(threadCount)
        val latch = CountDownLatch(threadCount)
        val barrier = CyclicBarrier(threadCount)

        for (i in 0..<threadCount) {
            executorService.submit {
                try {
                    barrier.await()
                    task.get()
                } catch (e: Exception) {
                    throw RuntimeException(e)
                } finally {
                    latch.countDown()
                }
            }
        }

        try {
            latch.await()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
        executorService.shutdown()
    }
}
